website: First draft of "Refactoring" documentation, about "moved" blocks (#29126)
hashicorp/terraform: website/docs/language/modules/develop/refactoring.html.md

---
layout: "language"
page_title: "Refactoring"
sidebar_current: "docs-modules-recactoring"
description: |-
  How to make backward-compatible changes to modules already in use.
---

# Refactoring

~> **Note:** Explicit refactoring declarations with `moved` blocks will work
only in Terraform v1.1 and later. When working with earlier Terraform versions,
or for refactoring actions too complex to express as `moved` blocks, you could
use
[the `terraform state mv` CLI command](/docs/cli/commands/state/mv.html)
as a separate step instead.

In shared modules and long-lived configurations, you may eventually outgrow
your initial module structure and resource names. For example, you might decide
that what was previously one child module makes more sense as two separate
modules and move a subset of the existing resources to the new one.

> Terraform compares previous state with new configuration, correlating by
> each module or resource's unique address. Therefore _by default_ Terraform will
> understand moving or renaming an object as an intent to destroy the object
> at the old address and to create a new object at the new address.

Reviewer: You write so well, Martin, that I'm nitpicking now :p But, I try to avoid saying "will." Often we can just make a stronger, simpler statement that means the same thing by just omitting it. 

```suggestion
Terraform compares previous state with new configuration, correlating by each module or resource's unique address. Therefore _by default_ Terraform understands moving or renaming an object as an intent to destroy the object at the old address and create a new object at the new address.
```

> When you add `moved` blocks in your configuration to record where you've
> historically moved or renamed an object, Terraform will instead treat an
> existing object at the old address as if it now belongs to the new address.

Reviewer: ```suggestion
When you add `moved` blocks in your configuration to record where you've
historically moved or renamed an object, Terraform treats an
existing object at the old address as if it now belongs to the new address.
```

## `moved` Block Syntax

A `moved` block expects no labels and contains only `from` and `to` arguments:

```hcl
moved {
  from = aws_instance.a
  to   = aws_instance.b
}
```

The example above records that the resource currently known as `aws_instance.b`
was known as `aws_instance.a` in a previous version of this module.

Before creating a new plan for `aws_instance.b`, Terraform first checks
whether there is an existing object for `aws_instance.a` recorded in the state.
If there is an existing object, Terraform renames that object to
`aws_instance.b` and then proceeds with creating a plan. The resulting plan is
as if the object had originally been created at `aws_instance.b`, avoiding any
need to destroy it during apply.

> The `from` and `to` addresses both use a special addressing syntax that allows
> selecting modules, resources, and resources inside child modules. Below, we
> describe the various different refactoring use-cases and the appropriate
> addressing syntax for each situation.

Reviewer: ```suggestion
The `from` and `to` addresses both use a special addressing syntax that allows
selecting modules, resources, and resources inside child modules. Below, we
describe several refactoring use-cases and the appropriate
addressing syntax for each situation.
```

* [Renaming a Resource](#renaming-a-resource)
* [Enabling `count` or `for_each` For a Resource](#enabling-count-or-for_each-for-a-resource)
* [Renaming a Module Call](#renaming-a-module-call)
* [Enabling `count` or `for_each` For a Module Call](#enabling-count-or-for_each-for-a-module-call)
* [Splitting One Module into Multiple](#splitting-one-module-into-multiple)
* [Removing `moved` blocks](#removing-moved-blocks)

## Renaming a Resource

> Imagine that the inital version of your module included a resource configuration
> like the following:

Reviewer: ```suggestion
Here is an example module with a resource configuration:
```

```hcl
resource "aws_instance" "a" {
  count = 2

  # (resource-type-specific configuration)
}
```

Applying this configuration for the first time would cause Terraform to
create `aws_instance.a[0]` and `aws_instance.a[1]`.

> If you later choose a better name for this resource, then you can change the
> name label in the `resource` block and record the old name inside a `moved` block:

Reviewer: ```suggestion
If you choose a different name for this resource in a subsequent module version, you can change the
name label in the `resource` block and record the old name inside a `moved` block:
```

```hcl
resource "aws_instance" "b" {
  count = 2

  # (resource-type-specific configuration)
}

moved {
  from = aws_instance.a
  to   = aws_instance.b
}
```

> When creating the next plan for each configuration using this module, Terraform
> will treat any existing objects belonging to `aws_instance.a` as if they had
> been created as `aws_instance.b`. `aws_instance.a[0]` will be treated as
> `aws_instance.b[0]`, and `aws_instance.a[1]` as `aws_instance.b[1]`.

Reviewer: ```suggestion
When creating the next plan for each configuration using this module, Terraform treats any existing objects belonging to `aws_instance.a` as if they had been created as `aws_instance.b`. `aws_instance.a[0]` will be treated as `aws_instance.b[0]`, and `aws_instance.a[1]` as `aws_instance.b[1]`.
```

New instances of the module, which _never_ had an
`aws_instance.a`, will ignore the `moved` block and propose to create
`aws_instance.b[0]` and `aws_instance.b[1]` as normal.

Both of the addresses in this example referred to a resource as a whole, and
so Terraform recognizes the move for all instances of the resource. That is,
it covers both `aws_instance.a[0]` and `aws_instance.a[1]` without the need
to identify each one separately.

Each resource type has a separate schema and so objects of different types
are not compatible. Therefore, although you can use `moved` to change the name
of a resource, you _cannot_ use `moved` to change to a different resource type
or to change a managed resource (a `resource` block) into a data resource
(a `data` block).

## Enabling `count` or `for_each` For a Resource

> Imagine that the initial version of your module contains a single-instance
> resource:

Reviewer: ```suggestion
Here is an example module that contains a single-instance
resource:
```

```hcl
resource "aws_instance" "a" {
  # (resource-type-specific configuration)
}
```

Applying this configuration would lead to Terraform creating an object
bound to the address `aws_instance.a`.

> Later, you need to use [`for_each`](../../meta-arguments/for_each.html) with this
> resource to systematically declare multiple instances. To preserve an object
> that was previously associated with `aws_instance.a` alone, you must use a
> `moved` block to specify which instance key the object will take in the new
> configuration:

Reviewer: ```suggestion
In a later version, you use [`for_each`](../../meta-arguments/for_each.html) with this
resource to systematically declare multiple instances. To preserve an object
that was previously associated with `aws_instance.a` alone, you must add a
`moved` block to specify which instance key the object will take in the new
configuration:
```

```hcl
locals {
  instances = tomap({
    big = {
      instance_type = "m3.large"
    }
    small = {
      instance_type = "t2.medium"
    }
  })
}

resource "aws_instance" "a" {
  for_each = local.instances

  instance_type = each.value.instance_type
  # (other resource-type-specific configuration)
}

moved {
  from = aws_instance.a
  to   = aws_instance.a["small"]
}
```

> The above will avoid Terraform planning to destroy any existing object at
> `aws_instance.a`, treating it instead as if it were originally created
> as `aws_instance.a["small"]`.

Reviewer: ```suggestion
The above will keep Terraform from planning to destroy any existing object at
`aws_instance.a`, treating those objects instead as if they were originally created
as `aws_instance.a["small"]`.
```

> When at least one of the two addresses includes an instance key, like
> `["small"]` in the above example, Terraform will understand both addresses

Reviewer: ```suggestion
When at least one of the two addresses includes an instance key, like
`["small"]` in the above example, Terraform understands both addresses
```

as referring to specific _instances_ of a resource rather than the resource
as a whole. That means you can use `moved` to switch between keys and to
add and remove keys as you switch between `count`, `for_each`, or neither.

The following are some other examples of valid `moved` blocks that record
changes to resource instance keys in a similar way:

```hcl
# Both old and new configuration used "for_each", but the
# "small" element was renamed to "tiny".
moved {
  from = aws_instance.b["small"]
  to   = aws_instance.b["tiny"]
}

# The old configuration used "count" and the new configuration
# uses "for_each", with the following mappings from
# index to key:
moved {
  from = aws_instance.c[0]
  to   = aws_instance.c["small"]
}
moved {
  from = aws_instance.c[1]
  to   = aws_instance.c["tiny"]
}

> # The old configuration used "count" and the new configuration
> # uses neither "count" nor "for_each", and we want to keep

Reviewer: ```suggestion
# The old configuration used "count" and the new configuration
# uses neither "count" nor "for_each", and you want to keep
```

module "a" {
  source = "../modules/example"

  # (module arguments)
}
```

When applying this configuration, Terraform would prefix the addresses for
any resources declared in this module with the module path `module.a`.
For example, a resource `aws_instance.example` would have the full address
`module.a.aws_instance.example`.

If you later choose a better name for this module call, then you can change the
name label in the `module` block and record the old name inside a `moved` block:

```hcl
module "b" {
  source = "../modules/example"

  # (module arguments)
}
```

When creating the next plan for each configuration using this module, Terraform
will treat any existing object addresses beginning with `module.a` as if
they had instead been created in `module.b`. `module.a.aws_instance.example`
would be treated as `module.b.aws_instance.example`.

Both of the addresses in this example referred to a module call as a whole, and
so Terraform recognizes the move for all instances of the call. If this
module call used `count` or `for_each` then it would apply to all of the
instances, without the need to specify each one separately.

## Enabling `count` or `for_each` For a Module Call

> Imagine that the initial version of your module contains a single-instance
> module:

Reviewer: ```suggestion
Here is an example of a single-instance module:
```

```hcl
module "a" {
  source = "../modules/example"

  # (module arguments)
}
```

Applying this configuration would cause Terraform to create objects whose
addresses begin with `module.a`.

> Later requirements might lead to you needing to use
> [`count`](../../meta-arguments/count.html) with this resource, in order
> to systematically declare multiple instances. To preserve an object that
> was previously associated with `aws_instance.a` alone, you can used a `moved`
> block to specify which instance key that object will take in the new
> configuration:

Reviewer: ```suggestion
In later module versions, you may need to use [`count`](../../meta-arguments/count.html) with this resource to systematically declare multiple instances. To preserve an object that was previously associated with `aws_instance.a` alone, you can add a `moved` block to specify which instance key that object will take in the new configuration:
```

```hcl
module "a" {
  source = "../modules/example"
  count  = 3

  # (module arguments)
}

moved {
  from = module.a
  to   = module.a[2]
}
```

> The above will cause Terraform to treat all objects in `module.a` as if they
> were originally created in `module.a[2]`, and thus it will plan to create
> new objects only for `module.a[0]` and `module.a[1]`.

Reviewer: Going to keep the "will" here as we're describing something that is truly future from the example. 

```suggestion
The configuration above directs Terraform to treat all objects in `module.a` as if they were originally created in `module.a[2]`. As a result, Terraform plans to create new objects only for `module.a[0]` and 'module.a[1]`.
```

> When at least one of the two addresses includes an instance key, like
> `[2]` in the above example, Terraform will understand both addresses as
> referring to specific _instances_ of a module call rather than the module
> call as a whole. That means you can use `moved` to switch between keys and to
> add and remove keys as you switch between `count`, `for_each`, or neither.

Reviewer: In general, we want to try to avoid clauses like this in the middle of the sentence. It's not _wrong_, it's just a bit harder to read than a more straightforward sentence construction. I'm removing this one here because I think no information is lost, but feel free to push back if you think it's necessary!

```suggestion
When at least one of the two addresses includes an instance key, Terraform will understand both addresses as
referring to specific _instances_ of a module call rather than the module
call as a whole. That means you can use `moved` to switch between keys and to
add and remove keys as you switch between `count`, `for_each`, or neither.
```

For more examples of recording moves associated with instances, refer to
the similar section
[Enabling `count` and `for_each` For a Resource](#enabling-count-or-for_each-for-a-resource).

# Splitting One Module into Multiple

As a module grows to support new requirements, it might eventually grow big
enough to warrant splitting into two separate modules.

> Consider the following initial version of a module:

Reviewer: ```suggestion
Consider this example module:
```

```hcl
resource "aws_instance" "a" {
  # (other resource-type-specific configuration)
}

resource "aws_instance" "b" {
  # (other resource-type-specific configuration)
}

resource "aws_instance" "c" {
  # (other resource-type-specific configuration)
}
```

You can split this into two modules as follows:

> * `aws_instance.a` will now belong to module "x".
> * `aws_instance.b` will also belong to module "x".
> * `aws_instance.c`, though, will belong to module "y".

Reviewer: ```suggestion
* `aws_instance.a` now belongs to module "x".
* `aws_instance.b` also belongs to module "x".
* `aws_instance.c` belongs to module "y".
```

To achieve this refactoring without replacing existing objects bound to the
old resource addresses, you must:

1. Write module "x", copying over the two resources it should contain.
2. Write module "y", copying over the one resource it should contain.
3. Edit the original module to no longer include any of these resources, and
   instead to contain only shim configuration to migrate existing users.

> The new modules "x" and "y" will therefore contain only `resource` blocks,
> as normal:

Reviewer: One last thought - you use "as normal" a couple of times I think and I'm not 100% sure what it means. I wonder if you can just omit it as I've suggested here. 

```suggestion
The new modules "x" and "y" should contain only `resource` blocks:
```

Author: With some time away from this _I'm_ honestly not sure what I meant by "as normal" here either, and I agree it doesn't really seem to be adding anything here.

(I _guess_ I was probably trying to say that there are no `moved` blocks in the child modules, because in this case they appear only in the root module, but another paragraph further down gets that message across more directly anyway. :man_shrugging: )

```hcl
# module "x"

resource "aws_instance" "a" {
  # (other resource-type-specific configuration)
}

resource "aws_instance" "b" {
  # (other resource-type-specific configuration)
}
```

```hcl
# module "y"

resource "aws_instance" "c" {
  # (other resource-type-specific configuration)
}
```

The original module, now only a shim for backward-compatibility, calls the
two new modules and indicates that the resources moved into them:

```hcl
module "x" {
  source = "../modules/x"

  # ...
}

module "y" {
  source = "../modules/y"

  # ...
}

moved {
  from = aws_instance.a
  to   = module.x.aws_instance.a
}

moved {
  from = aws_instance.b
  to   = module.x.aws_instance.b
}

moved {
  from = aws_instance.c
  to   = module.y.aws_instance.c
}
```

> When an existing user of the original module upgrades to the new "shim"
> version, Terraform will notice these three `moved` blocks and thus behave
> as if the objects associated with the three old resource addresses were
> originally created inside the two new modules.

Reviewer: ```suggestion
When an existing user of the original module upgrades to the new "shim"
version, Terraform notices these three `moved` blocks and behaves
as if the objects associated with the three old resource addresses were
originally created inside the two new modules.
```

New users of this family of modules may use either the combined shim module
_or_ the two new modules separately. You may wish to communicate to your
existing users that the old module is now deprecated and so they should use
the two separate modules for any new needs.

The multi-module refactoring situation is unusual in that it violates the
typical rule that a parent module sees its child module as a "closed box",
unaware of exactly which resources are declared inside it. This compromise
assumes that all three of these modules are maintained by the same people
and distributed together in a single
[module package](../sources.html#modules-in-package-sub-directories).

> To avoid strong [coupling](https://en.wikipedia.org/wiki/Coupling_(computer_programming))
> between separately-packaged modules, Terraform will only allow declarations of
> moves between modules in the same package. In other words, Terraform would
> not have allowed moving into `module.x` above if the `source` address of
> that call had not been a [local path](../sources.html#local-paths).

Reviewer: ```suggestion
To avoid strong [coupling](https://en.wikipedia.org/wiki/Coupling_(computer_programming))
between separately-packaged modules, Terraform only allows declarations of
moves between modules in the same package. In other words, Terraform would
not have allowed moving into `module.x` above if the `source` address of
that call had not been a [local path](../sources.html#local-paths).
```

Terraform resolves module references in `moved` blocks relative to the module
instance they are defined in. For example, if the original module above were
already a child module named `module.original`, the reference to
`module.x.aws_instance.a` would resolve as
`module.original.module.x.aws_instance.a`. A module may only make `moved`
statements about its own objects and objects of its child modules.

If you need to refer to resources within a module that was called using
`count` or `for_each` meta-arguments, you must specify a specific instance
key to use in order to match with the new location of the resource
configuration:

```hcl
moved {
  from = aws_instance.example
  to   = module.new[2].aws_instance.example
}
```

## Removing `moved` Blocks

Over time, a long-lasting module may accumulate many `moved` blocks.

> If you are maintaining private modules within an organization and you know
> for certain that all users of your module have successfully run
> `terraform apply` with your new module version already then it can be safe
> to remove `moved` blocks in later versions of your module.

Reviewer: Where possible, I try to rework sentences with a long "if" clause at the beginning. Flipping them around like this generally makes them easier to read :-) 

```suggestion
It can be safe to remove `moved` blocks in later versions of your module when you are maintaining private modules within an organization and all users of your module have successfully run
`terraform apply` with your new module version.
```

Author: This restructuring looks good to me. Thanks!

One way your new paragraph feels different to my original intent is that I was using "know for certain" (admittedly, clumsily) to try to suggest that this is something that the person making the decision needs to actively check, rather than just to _assume_.

Perhaps that's not really so important as I originally thought, but the consequence of getting this wrong is that a user of the module that _hadn't_ yet run `terraform apply` with a version including the `moved` block will see a plan to _delete_ the object that was moved, and applying that plan could be hard to recover from if the object is something like a database or some similar server which has state embedded in it.

I'm not super set on including this, but I wonder if you have some thoughts about this.

Author: When I was revising this I changed that part to say "...when you are maintaining private modules within an organization and _you know that_ all module users have...", but otherwise took this exactly as you wrote it.

I'm happy to still drop the extra words if you think they are unnecessary!

Reviewer: yeah that's fair - I actually went back and forth about that. It's minor and I think a little extra prompting never hurt anyone :-) I agree with what you did!

Author: Thanks!

However, removing a `moved` block is a generally breaking change
to your module because any configurations whose state refers to the old
address will then plan to delete that existing object instead of move it.

We recommend that by default module authors retain all historical `moved`
blocks from earlier versions of their modules, in order to preserve the
upgrade path for users of any old version. If later maintence causes you
to rename or move the same object twice, you can document that full history
using _chained_ `moved` blocks, where the new block refers to the existing
block:

```hcl
moved {
  from = aws_instance.a
  to   = aws_instance.b
}

moved {
  from = aws_instance.b
  to   = aws_instance.c
}
```

> Recording a sequence of moves in this way allows for successful upgrades for
> both configurations with objects at `aws_instance.a` _and_ configurations with
> objects at `aws_instance.b`. In both cases, Terraform will treat the existing
> object as if it had been originally created as `aws_instance.c`.

Reviewer: ```suggestion
Recording a sequence of moves in this way allows for successful upgrades for
both configurations with objects at `aws_instance.a` _and_ configurations with
objects at `aws_instance.b`. In both cases, Terraform treats the existing
object as if it had been originally created as `aws_instance.c`.
```

~> **Note:** Explicit refactoring declarations with `moved` blocks will work
only in Terraform v1.1 and later. When working with earlier Terraform versions,
or for refactoring actions too complex to express as `moved` blocks, you could
use
[the `terraform state mv` CLI command](/docs/cli/commands/state/mv.html)
as a separate step instead.

In shared modules and long-lived configurations you may eventually outgrow
your initial module heirarchy and resource naming and need to adopt a new
strategy. For example, you might decide that what was previously one
child module makes more sense as two separate modules, with each of the
resources moving to one of them.

Terraform compares previous state with new configuration, correlating by
each module or resource's unique address. Therefore _by default_ Terraform will
understand moving or renaming an object as an intent to destroy the object
at the old address and to create a new object at the new address.

You can use `moved` blocks in your configuration to record where you've
historically moved or renamed an object, in which case Terraform will instead
treat an existing object at the old address as if it now belongs to the new

> address.

Author: Good point!

I think the "why is this good" can vary between situations, so I'll have a think about how to write this in a more concise way, but in the interests of capturing the answer somewhere we can refer to it for later edits, reasons might include:

* As you suggested, it could be time-consuming to destroy and recreate some kinds of objects.
* The brief period where there are fewer (or no) objects of a particular type, or in some cases _too many_ objects of a particular type (if we're using `create_before_destroy`) can potentially cause total downtime or problematic backpressure on other subsystems. For example, adding new members to a Consul cluster will cause the cluster to temporarily thrash as it replicates data from the existing servers to the new servers, which might hurt performance for systems depending on that cluster.
* Some objects are not trivially replaceable. For example, if we have a resource representing a database server and we replace the database server then we'll probably be replacing a server which has data in it with a server that has _no_ data in it.

Perhaps the general way to talk about this is the idea that there are lots of infrastructure object types where replacing them requires careful planning and possibly coordination with other teams, and so module users and authors would typically prefer that _just_ upgrading to a new version of a module (without making other changes) would not involve replacing any existing objects.

I'll continue to ponder this, though if what I've said here gives _you_ any ideas about how to word this then of course I'd love to hear them!

> address.

Reviewer: I will think about this, too! Thank you as always for the awesome explanation :-)

## `moved` Block Syntax

A `moved` block expects no labels and contains only `from` and `to` arguments:

```hcl
moved {
  from = aws_instance.a
  to   = aws_instance.b
}
```

The example above records that the resource currently known as `aws_instance.b`
was, in an earlier version of this module, known instead as `aws_instance.a`.

Prior to creating a new plan for `aws_instance.b`, Terraform will first check
whether there's an existing object for `aws_instance.a` recorded in the state,
and if so will rename that object to `aws_instance.b` before taking any other
planning actions. The resulting plan is as if the object had originally been
created at `aws_instance.b`, avoiding any need to destroy that object.

The `from` and `to` addresses both use a special addressing syntax, different
than other references in a module, which allows selecting both modules and
resources, and possibly selecting resources inside child modules. The following
sections describe the various different refactoring use-cases, along with the

> appropriate addressing syntax for each situation.

Author: Agreed that the "possibly" doesn't really seem to be adding anything to this sentence, in retrospect. I think I was trying to say that the module path is an optional part of the address, but that's not really an important detail at this early point when we're just exploring the high-level structure.

The specific examples later hopefully serve to show the various different valid permutations.

* [Renaming a Resource](#renaming-a-resource)
* [Enabling `count` and `for_each` For a Resource](#enabling-count-and-for_each-for-a-resource)
* [Renaming a Module Call](#renaming-a-module-call)
* [Enabling `count` and `for_each` For a Module Call](#enabling-count-and-for_each-for-a-module-call)
* [Splitting One Module into Multiple](#splitting-one-module-into-multiple)
* [Removing `moved` blocks](#removing-moved-blocks)

## Renaming a Resource

> The following is an expanded version of the example above showing the full
> configurations at each step.

Reviewer: We probably don't need to say this - we can just launch in!

```suggestion
```

}
```

Applying this configuration for the first time would cause Terraform to
create `aws_instance.a[0]` and `aws_instance.a[1]`.

If you later choose a better name for this resource, then you can change the
name label in the `resource` block and record the old name inside a `moved` block:

```hcl
resource "aws_instance" "b" {
  count = 2

  # (resource-type-specific configuration)
}

moved {
  from = aws_instance.a
  to   = aws_instance.b
}
```

When creating the next plan for each configuration using this module, Terraform
will treat any existing objects belonging to `aws_instance.a` as if they had
been created as `aws_instance.b`. `aws_instance.a[0]` will be treated as
`aws_instance.b[0]`, and `aws_instance.a[1]` as `aws_instance.b[1]`.

New instances of the module, which _never_ had an
`aws_instance.a`, will ignore the `moved` block and propose to create
`aws_instance.b[0]` and `aws_instance.b[1]` as normal.

Both of the addresses in this example referred to a resource as a whole, and
so Terraform recognizes the move for all instances of the resource. That is,
it covers both `aws_instance.a[0]` and `aws_instance.a[1]` without the need
to identify each one separately.

> Each resource type has a separate schema and so objects of different types
> are not compatible. Therefore although you can use `moved` to change the name
> of a resource, you _cannot_ use `moved` to change to a different resource type
> or to change a managed resource (a `resource` block) into a data resource
> (a `data` block).

Reviewer: ```suggestion
Each resource type has a separate schema and so objects of different types
are not compatible. Therefore, although you can use `moved` to change the name
of a resource, you _cannot_ use `moved` to change to a different resource type
or to change a managed resource (a `resource` block) into a data resource
(a `data` block).
```

There is one exception to that constraint: a provider developer may mark a
particular resource type name as deprecated and nominate a new name to replace
it. In that case, the provider developer is responsible for making the old

> and new names schema-compatible, and can then tell Terraform that it's safe to

Reviewer: ```suggestion
and new names schema-compatible, and can then tell Terraform that it is safe to
```

support changing from the old to the new type via a `moved` block. Provider
documentation will include an example for any situation where this exception

> applies.

Reviewer: are we sure there will always be an example? I wonder if we should soften and say "should" or do we enforce this somehow?

Author: The mechanism I've described here doesn't exist yet today, so I wrote this with the intent that writing such documentation would be part of the work of making use of this new mechanism in providers.

However, it's true that this is written under the assumption that all such changes will be made by folks at HashiCorp who would therefore follow our documentation conventions/standards. We can't force third-party providers to have such documentation, so it is perhaps too strong to say "will".

With that said, there isn't really any way for a user to discover that this is possible _other than_ in documentation, so an undocumented instance of this would honestly be rather useless. I'm happy to change the wording as you suggested, but this is also a good note that we should make sure that the external provider-developer-facing documentation for this feature should mention this assumption too, once we have this facility available in the plugin SDK for provider developers to use. (This is actually not part of our initial scope of work here, so I could also just remove this statement altogether for now and restore it later once there's at least one resource type it's true for.)

Author: After re-reading the full section to recover the context, I decided to just remove this paragraph altogether for now.

We can add more words about this once this mechanism is implemented and so it'll presumably be clearer to us what provider developers will be expected to do in order to enable it.

## Enabling `count` or `for_each` For a Resource

Imagine that the initial version of your module contains a single-instance
resource:

```hcl
resource "aws_instance" "a" {
  # (resource-type-specific configuration)
}
```

Applying this configuration would lead to Terraform creating an object
bound to the address `aws_instance.a`.

> Later requirements might lead to you needing to use
> [`for_each`](../meta-arguments/for_each.html) with this resource, in order
> to systematically declare multiple instances. To preserve an object that

Reviewer: ```suggestion
Later, you need to use [`for_each`](../meta-arguments/for_each.html) with this resource to systematically declare multiple instances. To preserve an object that
```

`aws_instance.a`, treating it instead as if it were originally created
as `aws_instance.a["small"]`.

When at least one of the two addresses includes an instance key, like
`["small"]` in the above example, Terraform will understand both addresses
as referring to specific _instances_ of a resource rather than the resource
as a whole. That means you can use `moved` to switch between keys and to
add and remove keys as you switch between `count`, `for_each`, or neither.

The following are some other examples of valid `moved` blocks that record
changes to resource instance keys in a similar way:

```hcl
# Both old and new configuration used "for_each", but the
# "small" element was renamed to "tiny".
moved {
  from = aws_instance.b["small"]
  to   = aws_instance.b["tiny"]
}

# The old configuration used "count" and the new configuration
# uses "for_each", with the following mappings from
# index to key:
moved {
  from = aws_instance.c[0]
  to   = aws_instance.c["small"]
}
moved {
  from = aws_instance.c[1]
  to   = aws_instance.c["tiny"]
}

# The old configuration used "count" and the new configuration
# uses either "count" nor "for_each", and we want to keep

> # only the object at index 2.

Reviewer: should you change it to "or" rather than "nor"? OR should be be neither? 

Also, is there a way to rephrase without saying "we". Maybe we just change to "you"

Author: Hmm yes, I intended this to be "neither". Thanks!

Regarding the "we"/"you" thing, I think historically we've typically used "we" in situations where we're leading someone through a hypothetical example, and "you" for when we're talking about something the reader _actually_ wants, though I'm probably overstating how consistent we've been in that way.

For this particular situation it feels weird to me to _tell_ the reader what they want, but maybe that's okay. Another option that I think of considering your feedback here would be to phrase it as an "If you want ... then ..." sentence, perhaps like this:

```hcl
# The old configuration used "count" and the new configuration
# uses either "count" nor "for_each". If you want to keep the
# object at index 2, declare that it has moved to the new
# address.
moved {
  from = aws_instance.d[2]
  to   = aws_instance.d
}
```

Does that seem reasonable?

(I guess I'd also need to similarly edit the other examples above, because they are all currently phrased as descriptions of a situation where the example would apply, rather than as an instruction for the reader to follow, but I'll wait until I hear what you think about _this_ example before I attempt that! :grinning: )

were originally created in `module.a[2]`, and thus it will plan to create
new objects only for `module.a[0]` and `module.a[1]`.

When at least one of the two addresses includes an instance key, like
`[2]` in the above example, Terraform will understand both addresses as
referring to specific _instances_ of a module call rather than the module
call as a whole. That means you can use `moved` to switch between keys and to
add and remove keys as you switch between `count`, `for_each`, or neither.

For more examples of recording moves associated with instances, refer to
the similar section
[Enabling `count` and `for_each` For a Resource](#enabling-count-and-for_each-for-a-resource).

# Splitting One Module into Multiple

As a module grows to support new requirements, it might eventually grow big
enough to warrant splitting into two separate modules.

Consider the following initial version of a module:

```hcl
resource "aws_instance" "a" {
  # (other resource-type-specific configuration)
}

resource "aws_instance" "b" {
  # (other resource-type-specific configuration)
}

resource "aws_instance" "c" {
  # (other resource-type-specific configuration)
}
```

> Imagine that we intend to split this into two modules as follows:

Reviewer: ```suggestion
You can split this into two modules as follows:
```

Author: One thing I didn't really like about my first draft was the various places where I wrote "Imagine that...", or similar. That's not a sentence structure we typically use, but I was struggling for some other way to introduce a hypothetical situation before showing how to address that situation.

I know your feedback here is about this specific case, and this one is a little different than the other examples anyway, but I'd love to hear if you have any ideas on how to restructure the other "Imagine that..." sentences here.

Reviewer: Thank you for asking! I added in some additional comments to try to address this. Generally I think yes - we don't want to say "imagine" though I completely understand why you initially did that. It can be hard to do this without phrases like "Let's say" or "Imagine that". But, I think actually that it's just fine to say "Consider this example module:" [show example] "Later, you split this into two modules and add a block....."   I think this way you're making it active, concise, and still walking the user through the example in a way that makes sense. I think this also works because before this section you tell users that they're about to see a bunch of example walkthroughs. THoughts?

Author: Thanks! I've incorporated all of your suggestions below about "Imagine that..." sentences, and I previously incorporated the one I replied to here, so I'm going to resolve this conversation to represent that.

originally created inside the two new modules.

New users of this family of modules may use either the combined shim module
_or_ the two new modules separately. You may wish to communicate to your
existing users that the old module is now deprecated and so they should use
the two separate modules for any new needs.

The multi-module refactoring situation is unusual in that it violates the
typical rule that a parent module sees its child module as a "closed box",
unaware of exactly which resources are declared inside it. This compromise
assumes that all three of these modules are maintained by the same people
and distributed together in a single
[module package](sources.html#modules-in-package-sub-directories).

To avoid strong [coupling](https://en.wikipedia.org/wiki/Coupling_(computer_programming))
between separately-packaged modules, Terraform will only allow declarations of
moves between modules in the same package. In other words, Terraform would
not have allowed moving into `module.x` above if the `source` address of
that call had not been a [local path](sources.html#local-paths).

Module references in `moved` blocks are resolved relative to the module

> instance they are defined in. For example, if our original module above were

Reviewer: ```suggestion
instance they are defined in. For example, if the original module above were
```

already a child module named `module.original`, the reference to
`module.x.aws_instance.a` would resolve as
`module.original.module.x.aws_instance.a`. A module may only make `moved`
statements about its own objects and objects of its child modules.

If you need to refer to resources within a module that was called using
`count` or `for_each` meta-arguments, you must specify a specific instance
key to use in order to match with the new location of the resource
configuration:

```hcl
moved {
  from = aws_instance.example
  to   = module.new[2].aws_instance.example
}
```

## Removing `moved` Blocks

Over time, a long-lasting module may accumulate many `moved` blocks.

If you are maintaining private modules within an organization and you know
for certain that all users of your module have successfully run
`terraform apply` with your new module version already then it can be safe
to remove `moved` blocks in later versions of your module.

> However, in the general case removing a `moved` block is a breaking change
> to your module, because any configurations whose state refers to the old
> address will then plan to delete that existing object instead of move it.

Reviewer: ```suggestion
However, removing a `moved` block is a generally breaking change
to your module because any configurations whose state refers to the old
address will then plan to delete that existing object instead of move it.
```

We recommend that by default module authors retain all historical `moved`
blocks from earlier versions of their modules, in order to preserve the
upgrade path for users of any old version. If later maintence causes you
to rename or move the same object twice, you can document that full history
using _chained_ `moved` blocks, where the new block refers to the existing
block:

```hcl
moved {
  from = aws_instance.a
  to   = aws_instance.b
}

moved {
  from = aws_instance.b
  to   = aws_instance.c
}
```

> When you record a sequence of moves in this way you will allow for successful
> upgrades both for configurations with objects at `aws_instance.a` _and_
> configurations with objects at `aws_instance.b`. In both cases, Terraform will
> treat the existing object as if it had been originally created as
> `aws_instance.c`.

Reviewer: ```suggestion
Recording a sequence of moves in this way allows for successful
upgrades for both configurations with objects at `aws_instance.a` _and_
configurations with objects at `aws_instance.b`. In both cases, Terraform will
treat the existing object as if it had been originally created as
`aws_instance.c`.
```
